runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -761,27 +761,47 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
          }
 
       bool arrayRefGlobal;
+      bool storeValueGlobal;
+      const int storeValueOpIndex = isLoadFlattenableArrayElement ? -1 : 0;
       const int elementIndexOpIndex = isLoadFlattenableArrayElement ? 0 : 1;
       const int arrayRefOpIndex = elementIndexOpIndex+1;
 
       TR::Node *indexNode = node->getChild(elementIndexOpIndex);
       TR::Node *arrayRefNode = node->getChild(arrayRefOpIndex);
+      TR::Node *storeValueNode = isStoreFlattenableArrayElement ? node->getChild(storeValueOpIndex) : NULL;
       TR::VPConstraint *arrayConstraint = getConstraint(arrayRefNode, arrayRefGlobal);
       TR_YesNoMaybe isCompTypeVT = isArrayCompTypeValueType(arrayConstraint);
 
-      // If the array's component type is definitely not a value type, add a delayed
-      // transformation to replace the helper call with inline code to perform the
-      // array element access
+      TR_YesNoMaybe isStoreValueVT = isStoreFlattenableArrayElement ? isValue(getConstraint(storeValueNode, storeValueGlobal)) : TR_maybe;
+
+      // If the array's component type is definitely not a value type, or if the value
+      // being assigned in an array store operation is definitely not a value type, add
+      // a delayed transformation to replace the helper call with inline code to
+      // perform the array element access.
       //
-      if (arrayConstraint != NULL && isCompTypeVT == TR_no)
+      if ((arrayConstraint != NULL && isCompTypeVT == TR_no)
+          || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
          {
          flags8_t flagsForTransform(isLoadFlattenableArrayElement ? ValueTypesHelperCallTransform::IsArrayLoad
                                                                   : ValueTypesHelperCallTransform::IsArrayStore);
          flagsForTransform.set(ValueTypesHelperCallTransform::InsertDebugCounter);
 
          if (isStoreFlattenableArrayElement && !owningMethodDoesNotContainStoreChecks(this, node))
             {
-            flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
+            // If storing to an array whose component type is or might be a value type
+            // and the value that's being assigned is or might null, both a run-time
+            // NULLCHK of the value is required (guarded by a check of whether the
+            // component type is a value type) and an ArrayStoreCHK are required;
+            // otherwise, only the ArrayStoreCHK is required.
+            //
+            if ((isCompTypeVT != TR_no) && !storeValueNode->isNonNull())
```

Comment:
Should we just check the VP constraint on `storeValueNode` here rather than the node flag ?

Reply:
Good point.  Thanks!

Reply:
Fixed in commit 431245fe1fc19dd2646807f709c5769e46747765.  I can squash this commit when all reviews are completed.